### DISCUSSION TITLE
Implement HTML WebView in Task Properties window

### DISCUSSION THREAD

**Discussion by itzhouze** [imixs/open-bpmn#349] [Q&A]
Hello everyone,

We are currently working on a new project that will use OpenBPMN as a modeling platform for process diagrams.

We would now like to combine the form editor [@form-js](https://github.com/bpmn-io/form-js) with openBPMN to give the user the option of, for example, saving a simple form under a "UserTask" and displaying it in the PropertyWindow of the VisualStudioCode extension (using a WebView from [@form-js](https://github.com/bpmn-io/form-js)).

As I have already seen, it is possible to attach an "extensionElement" to a userTask.

```
    <bpmn2:userTask id="task_rpsmbQ" name="Some Task">
      <bpmn2:documentation id="documentation_eM7mzw" />
      <bpmn2:extensionElements>
        <hmp:Form Id="1" Name="Personal Data">
          <hmp:TextFieldDefinition Name="FirstName" Label="Vorname" Readonly="false" />
          <hmp:TextFieldDefinition Name="LastName" Label="Nachname" Readonly="false" />
          <hmp:TextFieldDefinition Name="NickName" Label="Nick-Name" Readonly="false" />
        </hmp:Form>
      </bpmn2:extensionElements>
      <bpmn2:incoming>sequenceFlow_C7VTFQ</bpmn2:incoming>
      <bpmn2:outgoing>sequenceFlow_uFjO0g</bpmn2:outgoing>
    </bpmn2:userTask>
```

My question now is the following:

How can we build an extension of OpenBPMN that allows us to include a [@form-js](https://github.com/bpmn-io/form-js) WebView in the form of a .html file in the VisualStudioCode properties window of a "UserTask, Task, ManualTask, ..."?

Unfortunately, I didn't get any further with the instructions: https://www.open-bpmn.org/glsp-server/BPMN_EXTENSIONS.html.

**Comment by itzhouze** [comment]
@rsoika @tortmayr - Any news here ? We would be very pleased to receive an answer :)

**Comment by rsoika** [comment]
Hi @itzhouze , sorry for the late reply - I have overseen your question.

Ok, there are two aspects: 
First your approach using the BPMN extension is correct and the [Open-BPMN meta project](https://github.com/imixs/open-bpmn/tree/master/open-bpmn.metamodel) allows you to operate on extensions and add you own.  This simply means BPMN 2.0 extensions can be handled by the open-bpmn meta project.

The other question you ask is how to integrate form-js into the client part. Of course this is possible. But at the moment I do not fully understand you requirement. Do you want that the Open-BPMN property panel displays your extension fields? Or do you want to model your form inside the property panel? 

As you may have notice we use [jsonforms.io](https://jsonforms.io/) to display the property panel. And we also provide a way that allows you to extend the properties displayed there. This can be controlled by the server part which means you can extend the property panel without changing the client side. You just need to implement a extension in Java for your new custom model server. For example we use this in [Imixs-Workflow](https://www.imixs.org). For example a element marked as a Imixs Element shows additional properties:

![grafik](https://github.com/imixs/open-bpmn/assets/1339124/04ab22e0-e91e-4708-bec6-124f9798c9ce)

Another solution is to simply connect a property (even a extension property) with an file inside your IDE. We use this often to describe more complex properties in a model: 

![grafik](https://github.com/imixs/open-bpmn/assets/1339124/58aee1c4-0623-4e7d-bfa7-f06c700e9d88)


This feature is independent and can be used also without changing the model via extensions.
